inflate_p.h

```diff
@@ -211,3 +211,81 @@ static inline void window_output_flush(PREFIX3(stream) *strm) {
     state->whave += out_bytes;
     state->whave = MIN(state->whave, state->wsize);
 }
+
+/* Behave like chunkcopy, but avoid writing beyond of legal output. */
```

Comment:
Hah, sort of in this case, since out overlaps len a lot.  Though I'm pretty sure the purpose of this function was not specifically to accommodate that scenario.

If you look at the code in chromium that a lot of this is based on, they too have several "relaxed", "safe", and "lapped" functions to accommodate all the different scenarios these things can be called and the potential data hazards.  In particular the "lapped" one is interesting.